## Projeto de Análise de Dados — Acidentes de Trânsito

### 1. Contextualização

Os acidentes de trânsito representam um problema relevante de segurança pública e podem gerar impactos sociais e econômicos. A análise de dados permite identificar padrões relacionados à ocorrência desses acidentes e contribuir para uma melhor compreensão do problema.

Neste projeto, serão analisados dados públicos da Polícia Rodoviária Federal (PRF) referentes aos acidentes registrados em rodovias federais brasileiras entre 2023 e 2025.

### 2. Objetivo

Identificar padrões, tendências e características relacionadas aos acidentes de trânsito, considerando aspectos como localização, período, causas, tipos e classificação das ocorrências.

Inicialmente, será realizada uma análise dos dados em âmbito nacional. A partir dos resultados encontrados, serão selecionados estados que apresentem volume e qualidade de dados adequados para uma análise mais aprofundada.

### 3. Fonte de dados

Os dados utilizados são provenientes da base de dados abertos da Polícia Rodoviária Federal (PRF), disponibilizada em arquivos CSV.

Foram utilizados os arquivos referentes aos anos de 2023, 2024 e 2025.

## 4. Tecnologias utilizadas

- Python
- Pandas
- SQL
- PySpark
- Power BI

### 5. Importação dos dados

Os arquivos CSV foram importados utilizando a biblioteca Pandas. Como os arquivos utilizam `;` como separador e a codificação `latin-1`, esses parâmetros foram especificados durante a leitura.

### 6. Importação dos dados

In [1]:
import pandas as pd

In [10]:
df_23 = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/projeto semantix/datatran2023.csv', encoding='latin-1', sep=';')
print(df_23.head())

       id data_inversa dia_semana   horario  uf   br     km  \
0  496519   2023-01-01    domingo  02:00:00  ES  101    114   
1  496543   2023-01-01    domingo  03:40:00  SP  116  113,1   
2  496590   2023-01-01    domingo  01:40:00  MT  163   1112   
3  496610   2023-01-01    domingo  10:40:00  PR  376  314,8   
4  496659   2023-01-01    domingo  14:55:00  MG  116  569,4   

           municipio                            causa_acidente  \
0          SOORETAMA            Ausência de reação do condutor   
1            TAUBATE             Entrada inopinada do pedestre   
2  GUARANTA DO NORTE  Reação tardia ou ineficiente do condutor   
3         ORTIGUEIRA                   Velocidade Incompatível   
4           MANHUACU         Acumulo de água sobre o pavimento   

               tipo_acidente  ... feridos_graves ilesos ignorados feridos  \
0  Saída de leito carroçável  ...              0      0         0       1   
1  Atropelamento de Pedestre  ...              0      0         4     

In [11]:
df_24 = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/projeto semantix/datatran2024.csv', encoding='latin-1', sep=';')
print(df_24.head())

         id data_inversa     dia_semana   horario  uf   br   km  \
0  571789.0   2024-01-01  segunda-feira  03:56:00  ES  101   38   
1  571804.0   2024-01-01  segunda-feira  04:50:00  PI  343  185   
2  571806.0   2024-01-01  segunda-feira  04:30:00  BA  116  578   
3  571818.0   2024-01-01  segunda-feira  06:30:00  SE  101   18   
4  571838.0   2024-01-01  segunda-feira  05:00:00  MT  364  240   

            municipio                                     causa_acidente  \
0  CONCEICAO DA BARRA                             Ultrapassagem Indevida   
1            PIRIPIRI                        Manobra de mudança de faixa   
2             BREJOES                   Ingestão de álcool pelo condutor   
3    MALHADA DOS BOIS           Reação tardia ou ineficiente do condutor   
4        RONDONOPOLIS  Condutor deixou de manter distância do veículo...   

                    tipo_acidente  ... feridos_graves ilesos ignorados  \
0  Colisão lateral sentido oposto  ...              1      1      

In [12]:
df_25 = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/projeto semantix/datatran2025.csv', encoding='latin-1', sep=';')
print(df_25.head())

       id data_inversa    dia_semana   horario  uf   br     km  \
0  652493   2025-01-01  quarta-feira  06:20:00  SP  116    225   
1  652519   2025-01-01  quarta-feira  07:50:00  CE  116  546,2   
2  652522   2025-01-01  quarta-feira  08:45:00  PR  369   88,2   
3  652544   2025-01-01  quarta-feira  11:00:00  PR  116     74   
4  652549   2025-01-01  quarta-feira  09:30:00  MG  251    471   

               municipio                            causa_acidente  \
0              GUARULHOS  Reação tardia ou ineficiente do condutor   
1              PENAFORTE                          Pista esburacada   
2      CORNELIO PROCOPIO  Reação tardia ou ineficiente do condutor   
3  CAMPINA GRANDE DO SUL  Reação tardia ou ineficiente do condutor   
4           FRANCISCO SA                   Velocidade Incompatível   

               tipo_acidente  ... feridos_graves ilesos ignorados feridos  \
0                 Tombamento  ...              0      0         1       1   
1            Colisão frontal


### 7. Conhecendo os dados

Nesta etapa será realizada uma análise inicial da estrutura das bases, verificando a quantidade de registros, o número de variáveis e os tipos de dados presentes em cada arquivo.

In [15]:
#7.1 Quantidade de registros e variáveis

print('df_23:',df_23.shape)
print('df_24:', df_24.shape)
print('df_25:', df_25.shape)

df_23: (67766, 30)
df_24: (73156, 30)
df_25: (72529, 30)


### 7.2 Variáveis disponíveis

A identificação das variáveis permite compreender quais informações estão disponíveis para a análise e verificar se a estrutura das bases permaneceu consistente entre os diferentes anos.

In [16]:
print('Colunas de 2023:')
print(df_23.columns.tolist())

print('\nColunas de 2024:')
print(df_24.columns.tolist())

print('\nColunas de 2025:')
print(df_25.columns.tolist())

Colunas de 2023:
['id', 'data_inversa', 'dia_semana', 'horario', 'uf', 'br', 'km', 'municipio', 'causa_acidente', 'tipo_acidente', 'classificacao_acidente', 'fase_dia', 'sentido_via', 'condicao_metereologica', 'tipo_pista', 'tracado_via', 'uso_solo', 'pessoas', 'mortos', 'feridos_leves', 'feridos_graves', 'ilesos', 'ignorados', 'feridos', 'veiculos', 'latitude', 'longitude', 'regional', 'delegacia', 'uop']

Colunas de 2024:
['id', 'data_inversa', 'dia_semana', 'horario', 'uf', 'br', 'km', 'municipio', 'causa_acidente', 'tipo_acidente', 'classificacao_acidente', 'fase_dia', 'sentido_via', 'condicao_metereologica', 'tipo_pista', 'tracado_via', 'uso_solo', 'pessoas', 'mortos', 'feridos_leves', 'feridos_graves', 'ilesos', 'ignorados', 'feridos', 'veiculos', 'latitude', 'longitude', 'regional', 'delegacia', 'uop']

Colunas de 2025:
['id', 'data_inversa', 'dia_semana', 'horario', 'uf', 'br', 'km', 'municipio', 'causa_acidente', 'tipo_acidente', 'classificacao_acidente', 'fase_dia', 'sentido_

### 7.3 Tipos de dados

Nesta etapa serão verificados os tipos de dados atribuídos às variáveis pelo Pandas. Essa verificação é importante para identificar possíveis inconsistências, como datas ou horários armazenados como texto, que poderão exigir tratamento posteriormente.

In [25]:
df_23.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67766 entries, 0 to 67765
Data columns (total 30 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   id                      67766 non-null  int64 
 1   data_inversa            67766 non-null  object
 2   dia_semana              67766 non-null  object
 3   horario                 67766 non-null  object
 4   uf                      67766 non-null  object
 5   br                      67766 non-null  int64 
 6   km                      67766 non-null  object
 7   municipio               67766 non-null  object
 8   causa_acidente          67766 non-null  object
 9   tipo_acidente           67766 non-null  object
 10  classificacao_acidente  67765 non-null  object
 11  fase_dia                67766 non-null  object
 12  sentido_via             67766 non-null  object
 13  condicao_metereologica  67766 non-null  object
 14  tipo_pista              67766 non-null  object
 15  tr

In [26]:
df_24.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73156 entries, 0 to 73155
Data columns (total 30 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      73156 non-null  float64
 1   data_inversa            73156 non-null  object 
 2   dia_semana              73156 non-null  object 
 3   horario                 73156 non-null  object 
 4   uf                      73156 non-null  object 
 5   br                      73156 non-null  int64  
 6   km                      73156 non-null  object 
 7   municipio               73156 non-null  object 
 8   causa_acidente          73156 non-null  object 
 9   tipo_acidente           73156 non-null  object 
 10  classificacao_acidente  73155 non-null  object 
 11  fase_dia                73156 non-null  object 
 12  sentido_via             73156 non-null  object 
 13  condicao_metereologica  73156 non-null  object 
 14  tipo_pista              73156 non-null

In [27]:

df_25.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72529 entries, 0 to 72528
Data columns (total 30 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   id                      72529 non-null  int64 
 1   data_inversa            72529 non-null  object
 2   dia_semana              72529 non-null  object
 3   horario                 72529 non-null  object
 4   uf                      72529 non-null  object
 5   br                      72529 non-null  int64 
 6   km                      72529 non-null  object
 7   municipio               72529 non-null  object
 8   causa_acidente          72529 non-null  object
 9   tipo_acidente           72529 non-null  object
 10  classificacao_acidente  72528 non-null  object
 11  fase_dia                72529 non-null  object
 12  sentido_via             72529 non-null  object
 13  condicao_metereologica  72529 non-null  object
 14  tipo_pista              72529 non-null  object
 15  tr

### 7.4 Considerações iniciais sobre a estrutura dos dados

As bases analisadas apresentam a mesma estrutura nos três anos, contendo 30 variáveis cada. O conjunto de dados possui 67.766 registros em 2023, 73.156 em 2024 e 72.529 em 2025, totalizando 213.451 registros.

A padronização das colunas entre os anos permite a posterior consolidação das bases para realização de análises comparativas e temporais.

Durante a análise inicial, foram identificadas algumas diferenças nos tipos de dados entre os arquivos. A variável `id`, por exemplo, está armazenada como inteiro nos arquivos de 2023 e 2025, mas como número decimal em 2024. Também foram observadas diferenças nos tipos das variáveis `latitude` e `longitude`, que estão armazenadas como texto em alguns anos e como números em outro.

Além disso, foram identificados poucos valores ausentes em algumas variáveis, principalmente `classificacao_acidente`, `regional`, `delegacia` e `uop`. A quantidade e a distribuição desses valores serão investigadas na etapa de qualidade dos dados antes da definição das estratégias de tratamento.

## 8. Qualidade dos dados

Nesta etapa serão avaliadas possíveis inconsistências presentes nas bases, incluindo valores ausentes, registros duplicados e diferenças de formato ou tipo das variáveis.

A identificação desses problemas é necessária para garantir maior consistência nas análises realizadas posteriormente.

### 8.1 Valores ausentes

Inicialmente, será verificada a quantidade de valores ausentes em cada variável dos três conjuntos de dados.

In [35]:
print('Valores ausentes - 2023')
print(df_23.isnull().sum())

Valores ausentes - 2023
id                         0
data_inversa               0
dia_semana                 0
horario                    0
uf                         0
br                         0
km                         0
municipio                  0
causa_acidente             0
tipo_acidente              0
classificacao_acidente     1
fase_dia                   0
sentido_via                0
condicao_metereologica     0
tipo_pista                 0
tracado_via                0
uso_solo                   0
pessoas                    0
mortos                     0
feridos_leves              0
feridos_graves             0
ilesos                     0
ignorados                  0
feridos                    0
veiculos                   0
latitude                   0
longitude                  0
regional                  17
delegacia                 32
uop                       46
dtype: int64


In [36]:
print('\nValores ausentes - 2024')
print(df_24.isnull().sum())


Valores ausentes - 2024
id                          0
data_inversa                0
dia_semana                  0
horario                     0
uf                          0
br                          0
km                          0
municipio                   0
causa_acidente              0
tipo_acidente               0
classificacao_acidente      1
fase_dia                    0
sentido_via                 0
condicao_metereologica      0
tipo_pista                  0
tracado_via                 0
uso_solo                    0
pessoas                     0
mortos                      0
feridos_leves               0
feridos_graves              0
ilesos                      0
ignorados                   0
feridos                     0
veiculos                    0
latitude                    0
longitude                   0
regional                    3
delegacia                  36
uop                       112
dtype: int64


In [37]:
print('\nValores ausentes - 2025')
print(df_25.isnull().sum())


Valores ausentes - 2025
id                         0
data_inversa               0
dia_semana                 0
horario                    0
uf                         0
br                         0
km                         0
municipio                  0
causa_acidente             0
tipo_acidente              0
classificacao_acidente     1
fase_dia                   0
sentido_via                0
condicao_metereologica     0
tipo_pista                 0
tracado_via                0
uso_solo                   0
pessoas                    0
mortos                     0
feridos_leves              0
feridos_graves             0
ilesos                     0
ignorados                  0
feridos                    0
veiculos                   0
latitude                   0
longitude                  0
regional                   2
delegacia                 22
uop                       38
dtype: int64


### 8.1.1 Análise dos valores ausentes

A análise dos valores ausentes mostrou que as bases apresentam elevado nível de completude. As variáveis relacionadas às características dos acidentes, como `uf`, `causa_acidente`, `tipo_acidente`, `classificacao_acidente`, `pessoas`, `mortos` e `feridos`, apresentam praticamente todos os registros preenchidos.

Os valores ausentes estão concentrados principalmente nas variáveis `regional`, `delegacia` e `uop`, que possuem caráter administrativo e operacional. A maior quantidade identificada foi de 112 registros ausentes na variável `uop` em 2024, correspondendo a uma pequena parcela do total de registros daquele ano.

Diante da baixa quantidade de valores ausentes, não será realizada a exclusão dos registros nesta etapa. Os dados serão mantidos na base e, quando necessário, os valores ausentes serão tratados de acordo com a variável e com a finalidade de cada análise.

### 8.1.2 Distribuição dos registros com valores ausentes

Será verificado se os valores ausentes estão concentrados nos mesmos registros ou distribuídos entre diferentes ocorrências.

In [39]:
print('Registros com pelo menos um valor ausente:')
print('2023:', df_23.isnull().any(axis=1).sum())
print('2024:',df_24.isnull().any(axis=1).sum())
print('2025:', df_25.isnull().any(axis=1).sum())

print('Percentual de registros com pelo menos um valor ausente:')
print('2023:', round(df_23.isnull().any(axis=1).mean() * 100,3),'%')
print('2024:', round(df_24.isnull().any(axis=1).mean() * 100,3), '%')
print('2025:', round(df_25.isnull().any(axis=1).mean() * 100,3), '%')

Registros com pelo menos um valor ausente:
2023: 47
2024: 113
2025: 39
Percentual de registros com pelo menos um valor ausente:
2023: 0.069 %
2024: 0.154 %
2025: 0.054 %


Os registros que apresentam pelo menos um valor ausente representam uma parcela muito pequena das respectivas bases: 0,069% em 2023, 0,154% em 2024 e 0,054% em 2025.

Dessa forma, os dados apresentam elevado nível de completude, não sendo necessária a exclusão dos registros por ausência de informações. Os registros serão preservados e os valores ausentes poderão ser tratados pontualmente caso alguma análise específica utilize as variáveis afetadas.

### 8.2 Registros duplicados

Nesta etapa será verificada a existência de registros duplicados nas bases. Inicialmente, serão identificadas linhas completamente iguais, ou seja, registros em que todas as variáveis apresentam exatamente os mesmos valores.

In [42]:
print('Duplicidades - 2023')
print(df_23.duplicated().sum())

print('\nDuplicidade - 2024')
print(df_24.duplicated().sum())

print('\nDuplicidade - 2025')
print(df_25.duplicated().sum())

print('\nId duplicado 2023')
print(df_23['id'].duplicated().sum())

print('\nId duplicado 2024')
print(df_24['id'].duplicated().sum())

print('\nId duplicado 2025')
print(df_25['id'].duplicated().sum())


Duplicidades - 2023
0

Duplicidade - 2024
0

Duplicidade - 2025
0

Id duplicado 2023
0

Id duplicado 2024
0

Id duplicado 2025
0


### 8.2.1 Resultado da análise de duplicidades

A análise não identificou registros completamente duplicados em nenhum dos três conjuntos de dados. Também não foram encontrados identificadores (`id`) duplicados dentro das bases de 2023, 2024 e 2025.

Dessa forma, não foi identificada a necessidade de remoção de registros por duplicidade nesta etapa.

### 8.3 Verificação das variáveis numéricas

Durante a análise da estrutura dos dados foram identificadas diferenças nos tipos atribuídos a algumas variáveis entre os anos. Nesta etapa serão investigados os valores presentes nessas colunas para determinar se será necessário realizar uma padronização antes da análise.

In [53]:
print('Tipo de ID')

print('2023:', df_23.dtypes['id'])
print('2024:', df_24.dtypes['id'])
print('2025:', df_25.dtypes['id'])

print('\nValores de ID com casas decimais em 2024:')
print(df_24['id'].head(10))

# latitude e longitude

print('\nTipo de Latitude')
print('2023:', df_23['latitude'].head(5).tolist())
print('2024:', df_24['latitude'].head(5).tolist())
print('2025:', df_25['latitude'].head(5).tolist())

print('\nTipo de Longitude')
print('2023:', df_23['longitude'].head(5).tolist())
print('2024:', df_24['longitude'].head(5).tolist())
print('2025:', df_25['longitude'].head(5).tolist())

print('\nExemplo de KM:')
print('2023:', df_23['km'].head(10).tolist())
print('2024:', df_24['km'].head(10).tolist())
print('2025:', df_25['km'].head(10).tolist())

Tipo de ID
2023: int64
2024: float64
2025: int64

Valores de ID com casas decimais em 2024:
0    571789.0
1    571804.0
2    571806.0
3    571818.0
4    571838.0
5    571855.0
6    571910.0
7    571922.0
8    571949.0
9    571957.0
Name: id, dtype: float64

Tipo de Latitude
2023: ['-19,09484877', '-23,0445658', '-9,70020602', '-23,985512', '-20,10007457']
2024: [-18.48261, -4.29603281, -13.07158302, -10.35601949, -16.17914141]
2025: ['-23,48586772', '-7,812288', '-23,182565', '-25,36517687', '-16,46801304']

Tipo de Longitude
2023: ['-40,05095848', '-45,58259814', '-54,87588757', '-51,083555', '-42,17884091']
2024: [-39.92379, -41.76732659, -39.9611107, -36.90552235, -54.78905337]
2025: ['-46,54075317', '-39,08333306', '-50,637228', '-49,04223028', '-43,43121303']

Exemplo de KM:
2023: ['114', '113,1', '1112', '314,8', '569,4', '569,8', '152', '897,3', '105,5', '366']
2024: ['38', '185', '578', '18', '240', '447', '852,9', '459,7', '550', '594,4']
2025: ['225', '546,2', '88,2', '74', '

In [56]:
# Segunda verificação antes das mudanças

print('--- ID 2024 ---')
print('Quantidade ID\'s não inteiros')
print((df_24['id'] % 1 != 0). sum())

print('\n --- LATITUDE ---')
print('Valores Nulos:')
print(
    df_23['longitude'].isnull().sum(),
    df_24['longitude'].isnull().sum(),
    df_25['longitude'].isnull().sum()
)

print('\n --- KM ---')
print('Valores Nulos:')
print(
    df_23['km'].isnull().sum(),
    df_24['km'].isnull().sum(),
    df_25['km'].isnull().sum()
)

# Verificação se existem valores estranhos como 'km50' 'não informado'

print('\nValores únicos não numéricos em KM:')

for ano, df in [('2023', df_23), ('2024', df_24), ('2025', df_25)]:
    km_teste = pd.to_numeric(
        df['km'].astype(str).str.replace(',', '.', regex=False),
        errors='coerce'
    )

    print(f'\n{ano}:')
    print(df.loc[km_teste.isna(), 'km'].value_counts())

--- ID 2024 ---
Quantidade ID's não inteiros
0

 --- LATITUDE ---
Valores Nulos:
0 0 0

 --- KM ---
Valores Nulos:
0 0 0

Valores únicos não numéricos em KM:

2023:
Series([], Name: count, dtype: int64)

2024:
Series([], Name: count, dtype: int64)

2025:
Series([], Name: count, dtype: int64)


### 8.3.1 Resultado da Verificação das variáveis numéricas

Foram analisadas as variáveis id, latitude, longitude e km, considerando seus tipos de dados, valores nulos e formatos de representação.

A variável id apresentou tipo float64 apenas em 2024, enquanto nos demais anos foi carregada como int64. A verificação mostrou que nenhum identificador possui parte decimal, indicando uma diferença de representação dos dados, e não uma característica dos valores.

As variáveis latitude e longitude não apresentaram valores nulos. Entretanto, nos arquivos de 2023 e 2025 os valores utilizam vírgula como separador decimal e foram carregados como texto, enquanto em 2024 já estão em formato numérico.

A variável km também não apresentou valores nulos. Apesar de estar armazenada como texto nos três anos, os valores encontrados possuem representação numérica, utilizando vírgula como separador decimal em parte dos registros.

Dessa forma, será necessário padronizar os formatos dessas variáveis antes da consolidação dos dados, garantindo que os três anos utilizem os mesmos tipos de dados.

### 8.3.1 — Padronização do id

In [57]:
from IPython.core.shellapp import pt_inputhooks
df_23['id'] = df_23['id'].astype('int64')
df_24['id'] = df_24['id'].astype('int64')
df_25['id'] = df_25['id'].astype('int64')

print('Tipo de Dados')
print('2023:', df_23['id'].dtype)
print('2024:', df_24['id'].dtype)
print('2025:', df_25['id'].dtype)

Tipo de Dados
2023: int64
2024: int64
2025: int64


### 8.3.3 — Padronização de latitude e longitude

In [58]:
# Padronização da latitude

df_23['latitude'] = pd.to_numeric(
    df_23['latitude'].str.replace(',', '.', regex=False)
)

df_24['latitude'] = pd.to_numeric(df_24['latitude'])

df_25['latitude'] = pd.to_numeric(
    df_25['latitude'].str.replace(',', '.', regex=False)
)

# Padronização da longitude

df_23['longitude'] = pd.to_numeric(
    df_23['longitude'].str.replace(',', '.', regex=False)
)

df_24['longitude'] = pd.to_numeric(df_24['longitude'])

df_25['longitude'] = pd.to_numeric(
    df_25['longitude'].str.replace(',', '.', regex=False)
)

print('\nTipos após tratamento:')
print('--- Latitude ---')
print('2023:', df_23['latitude'].dtype)
print('2024:', df_24['latitude'].dtype)
print('2025:', df_25['latitude'].dtype)
print('\n--- Longitude ---')
print('2023:', df_23['longitude'].dtype)
print('2024:', df_24['longitude'].dtype)
print('2025:', df_25['longitude'].dtype)


Tipos após tratamento:
--- Latitude ---
2023: float64
2024: float64
2025: float64

--- Longitude ---
2023: float64
2024: float64
2025: float64
